### PR TITLE
feat(core): complete define_metric feature parity

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------

- Fixes WB-17821

This PR implements the majority of the missing `define_metric` functionality, most notably the different summary aggregations.

Exceptions:
- Dealing with NaNs when computing mean/min/max summary values. We should instead add nan<metric> specs like numpy.
- Dotted notation / nested summary value behavior.
- ~Something funky going on with config/telemetry~

The first two seem low priority, will fix in a follow up PR.

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

### CODE REVIEW

File: core/internal/filestream/updatesummary.go [hunk 1]
```diff
@@ -14,7 +14,9 @@ type SummaryUpdate struct {
 }
 
 func (u *SummaryUpdate) Apply(ctx UpdateContext) error {
-	rs := runsummary.New()
+	// fmt.Println("filestream: SummaryUpdate.Apply")
+	// fmt.Printf("filestream: SummaryUpdate.Apply: %v\n", *u)
```
Comment (dmitryduev) [inline core/internal/filestream/updatesummary.go:18]:
```suggestion
```

Comment (kptkin) [inline core/internal/filestream/updatesummary.go:18]:
```suggestion
```

File: core/internal/runmetric/runmetric.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	// for pattern, globMetric := range mh.globMetrics {
+	// 	fmt.Printf("    pattern: %v, globMetric: %v\n", pattern, *globMetric)
+	// }
+	// for pattern, definedMetric := range mh.definedMetrics {
+	// 	fmt.Printf("    pattern: %v, definedMetric: %v\n", pattern, *definedMetric)
+	// }
+
+	for pattern, globMetric := range mh.GlobMetrics {
+		if match, err := filepath.Match(pattern, key); err != nil {
+			// h.logger.CaptureError("error matching metric", err)
```
Comment (kptkin) [inline core/internal/runmetric/runmetric.go:37]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 3]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+func (rs *RunSummary) GetSummaryTypes(path []string) []SummaryType {
+	// look for a matching rule
+	// TODO: properly implement dot notation for nested keys,
+	// see test_metric_full.py::test_metric_dotted for an example
+	name := strings.Join(path, ".")
+
+	types := make([]SummaryType, 0)
+
+	for pattern, definedMetric := range rs.mh.DefinedMetrics {
+		// fmt.Printf("    kp: %v, pattern: %v\n", kp, pattern)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:76]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 4]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+		if pattern == name {
+			summary := definedMetric.GetSummary()
+			// fmt.Printf("    found defined metric. summary: %v\n", summary)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:79]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 5]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+			if summary.GetNone() {
+				// fmt.Printf("+++    requested none summary for metric %v\n", name)
+				return []SummaryType{None}
+			}
+			if summary.GetMax() {
+				// fmt.Printf("+++    requested max summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:85]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 6]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+			if summary.GetMin() {
+				// fmt.Printf("+++    requested min summary for metric %v\n", name)
+				types = append(types, Min)
+			}
+			if summary.GetMean() {
+				// fmt.Printf("+++    requested mean summary for metric %v\n", name)
+				types = append(types, Mean)
+			}
+			if summary.GetLast() {
+				// fmt.Printf("+++    requested last summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:97]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 7]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+		}
+	}
+	for pattern, globMetric := range rs.mh.GlobMetrics {
+		// fmt.Printf("    kp: %v, pattern: %v\n", kp, pattern)
+		// match the key against the glob pattern:
+		if match, err := filepath.Match(pattern, name); err == nil {
+			if match {
+				summary := globMetric.GetSummary()
+				if summary.GetNone() {
+					// fmt.Printf("---    requested none summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:109]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 8]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+					return []SummaryType{None}
+				}
+				// fmt.Printf("    found glob metric. summary: %v\n", summary)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:112]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 9]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+				if summary.GetMax() {
+					// fmt.Printf("---    requested max summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:114]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 10]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+					types = append(types, Max)
+				}
+				if summary.GetMin() {
+					// fmt.Printf("---    requested min summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:118]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 11]
```diff
@@ -2,25 +2,135 @@ package runsummary
...
+					types = append(types, Min)
+				}
+				if summary.GetMean() {
+					// fmt.Printf("---    requested mean summary for metric %v\n", name)
+					types = append(types, Mean)
+				}
+				if summary.GetLast() {
+					// fmt.Printf("---    requested last summary for metric %v\n", name)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:126]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 12]
```diff
@@ -31,26 +141,89 @@ func (rs *RunSummary) ApplyChangeRecord(
 	summaryRecord *service.SummaryRecord,
 	onError func(error),
 ) {
-
+	// handle updates
 	updates := make([]*pathtree.PathItem, 0, len(summaryRecord.GetUpdate()))
+
 	for _, item := range summaryRecord.GetUpdate() {
+		// fmt.Printf(">>> key: %v\n", item.GetKey())
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:148]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 13]
```diff
@@ -31,26 +141,89 @@ func (rs *RunSummary) ApplyChangeRecord(
...
+		// update all the stats for the given key path
+		kp := keyPath(item)
+		err = rs.stats.UpdateStats(kp, update)
+		if err != nil {
+			onError(err)
+			continue
+		}
+		// get the summary type for the item
+		// fmt.Printf("      mh: %v\n", *rs.mh)
+		// fmt.Printf("      update: %v\n", update)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:163]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 14]
```diff
@@ -31,26 +141,89 @@ func (rs *RunSummary) ApplyChangeRecord(
...
+
+		st := rs.GetSummaryTypes(kp)
+
+		// skip if None in the summary type slice
+		if len(st) == 1 && st[0] == None {
+			continue
+		}
+
+		// get the requested stats for the item
+		// fmt.Printf("+++    requested stat for metric %v: %v\n", kp, st)
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:173]:
```suggestion
```

File: core/pkg/server/handler.go [hunk 15]
```diff
@@ -395,16 +396,17 @@ func (h *Handler) handleStepMetric(key string) {
 func (h *Handler) handleMetric(record *service.Record, metric *service.MetricRecord) {
 	// metric can have a glob name or a name
 	// TODO: replace glob-name/name with one-of field
+	// fmt.Printf("metric: %v\n", metric)
```
Comment (kptkin) [inline core/pkg/server/handler.go:399]:
```suggestion
```

File: core/pkg/server/handler.go [hunk 16]
```diff
@@ -1312,9 +1316,11 @@ func (h *Handler) matchHistoryItemMetric(item *service.HistoryItem) *service.Met
 func (h *Handler) imputeStepMetric(item *service.HistoryItem) *service.HistoryItem {
 
 	// check if history item matches a defined metric or a glob metric
+	// fmt.Printf("item: %v\n", item)
```
Comment (kptkin) [inline core/pkg/server/handler.go:1319]:
```suggestion
```

File: core/pkg/server/handler.go [hunk 17]
```diff
@@ -1312,9 +1316,11 @@ func (h *Handler) matchHistoryItemMetric(item *service.HistoryItem) *service.Met
...
 	metric := h.matchHistoryItemMetric(item)
 
 	key := metric.GetStepMetric()
+	// fmt.Printf(" key: %v, metric: %v\n", key, metric)
```
Comment (kptkin) [inline core/pkg/server/handler.go:1323]:
```suggestion
```

File: core/pkg/server/handler.go [hunk 18]
```diff
@@ -1330,6 +1336,7 @@ func (h *Handler) imputeStepMetric(item *service.HistoryItem) *service.HistoryIt
 	// TODO: avoid using json marshalling
 	// we use the summary value of the metric as the algorithm for imputing the step metric
 	if value, ok := h.runSummary.Tree()[key]; ok {
+		// fmt.Printf("  value: %v\n", value)
```
Comment (kptkin) [inline core/pkg/server/handler.go:1339]:
```suggestion
```

File: core/internal/runsummary/runsummary.go [hunk 19]
```diff
@@ -2,25 +2,121 @@ package runsummary
...
 func NewFrom(tree pathtree.TreeData) *RunSummary {
-	return &RunSummary{pathTree: pathtree.NewFrom(tree)}
+	return &RunSummary{
+		pathTree: pathtree.NewFrom(tree),
+		stats:    statsTreeFromPathTree(tree),
+		mh:       runmetric.NewMetricHandler(),
+	}
+}
+
+// GetSummary matches the path against the defined metrics and returns the
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:62]:
```suggestion
// GetSummaryType matches the path against the defined metrics and returns the
```

File: core/internal/runsummary/runsummary.go [hunk 20]
```diff
@@ -2,25 +2,121 @@ package runsummary
...
+				types = append(types, Min)
+			}
+			if summary.GetMean() {
+				types = append(types, Mean)
+			}
+			if summary.GetLast() {
+				types = append(types, Latest)
+			}
+		}
+	}
```
Comment (kptkin) [inline core/internal/runsummary/runsummary.go:94]: how about using switch case here?

**Review by kptkin — APPROVED** [review]

**Pull request merged**